backend/core/models/user.py

```diff
@@ -227,21 +227,78 @@ def trusted_users(cls) -> QuerySet["User"]:
     def supertrusted_seed_users(cls) -> QuerySet["User"]:
         return cls.trusted_users().filter(is_staff=True)
 
+    @classmethod
+    def validate_email_unique_with_plus(cls, email: str, username="") -> str:
+        """Raise ValidationError when similar emails are found in the database.
+
+        Keyword arguments:
+        email -- An email that is going to be written in the database.
+        username -- The logged user's username, used to exclude him/herself from
+                    the validation when updating its own email. Empty for
+                    anonymous users.
+
+        Emails considered similar when:
+            - they share the same non-case-sensitive domain ;
+            - they share, in the local part, the same non case-sensitive string
+              before the `+` symbol.
+
+        Examples of emails considered similar:
+            - bob@example.org
+            - bob+@example.org
+            - bob+tournesol@example.org
+            - BOB+tournesol@example.org
+            - bob+hello@example.org
+            - BOB+HELLO@example.org
+            - etc.
+        """
+        email_split = email.split("@")
+
+        # if there is no `@`, do nothing
+        if len(email_split) == 1:
+            return email
+
+        local_part = email_split[0]
+        local_part_split = local_part.split("+")
+
+        users = User.objects.filter(
+            Q(email__iexact=f"{local_part_split[0]}@{email_split[-1]}")
+            | (
+                    Q(email__istartswith=f"{local_part_split[0]}+")
+                    & Q(email__iendswith=f"@{email_split[-1]}")
+            ),
+        )
+
+        if username:
+            users.exclude(username=username)
+
+        if users.exists():
+            raise ValidationError(
+                {
+                    "email": _(
+                        "A user with an email starting with '%(email)s' already exists"
+                        " in this domain." % {"email": f"{local_part_split[0]}+"}
+                    )
+                }
+            )
```

Comment:
After verification, using the Django `ValidationError` is fine and is correctlly handled by DRF to format the error, and make it correctly rendered on frontend :+1:

It's also possible to call this validation in `validate_email` instead of `validate`. In this case the ValidationError needs to receive the error string only (without the "email" key).

Reply:
Great!

Removing the `email` key is fine, but as the email validation is also called in `User.clean`, the error message won't be displayed near the email field in the administration interface.

To keep the message near the email field, it will required either **(1)** to keep the validation in the general `validate` method of the serializers ; or **(2)** to move the exception raising outside of the `validate_email_unique_with_plus` method, so that each parent methods will be able to format the exception correctly ; or **(3)** or keep it as is, remove the key `email` in the exception, and from the `User.email` method catch and re-raise an exception to add the missing email field.

Reply:
I would find option **`(3)`** slightly more explicit, but either (1) or (3) looks good to me.

Reply:
I implemented **(3)**.

Fixed by https://github.com/tournesol-app/tournesol/pull/825/commits/5b07b80a1a8c0b08d2b97e3a498207786bba733e.